stacks-signer/src/runloop.rs

```diff
@@ -797,9 +800,151 @@ impl<C: Coordinator> SignerRunLoop<Vec<OperationResult>, RunLoopCommand> for Run
 }
 
 /// Helper function for determining the coordinator public key given the the public keys
-fn calculate_coordinator(public_keys: &PublicKeys) -> (u32, ecdsa::PublicKey) {
-    // TODO: do some sort of VRF here to calculate the public key
-    // See: https://github.com/stacks-network/stacks-blockchain/issues/3915
-    // Mockamato just uses the first signer_id as the coordinator for now
-    (0, public_keys.signers.get(&0).cloned().unwrap())
+pub fn calculate_coordinator(
+    public_keys: &PublicKeys,
+    stacks_client: &StacksClient,
+) -> (u32, ecdsa::PublicKey) {
+    let stacks_tip_consensus_hash = match stacks_client.get_stacks_tip_consensus_hash() {
```

Comment:
There are a few of interesting consequences to using the Stacks chain tip consensus hash as the "seed" for choosing a coordinator at random.  They all stem from the fact that the consensus hash only changes whenever the node discovers and processes a new Bitcoin block.

* Stacks nodes A and B both process the same history Bitcoin blocks, and both happen to call `calculate_coordinator` _after_ they have processed the same history (this is the happy path)
* Stacks node A has not processed the full history of Bitcoin blocks (e.g. it's behind by one block), while B has. In this case, A will initially choose the wrong coordinator, but it will eventually choose the right coordinator once its node has caught up.
* Stacks nodes A and B each process divergent histories of Bitcoin blocks (i.e. different forks of Bitcoin), which cause A and B to calculate different coordinators for the same Bitcoin block.

The latter two cases are rare but they do happen in practice (the 2nd case is far more common than the 3rd case).  Regardless of which happens, node A's signer will choose the wrong coordinator.  How does node A recover?

Reply:
As far as I am aware, the system will currently just ignore that signer. it will not delcare it byzantine. We filter out "non coordinator' coordinator messages from the signer runloop with zero consequence to that signer.

Reply:
>  I think modifying the messages sent through stacker db to include additional info such as the consensus tip and the signers responding with the NACK with hints that someone has an incorrect view of chain tip would be a good improvement ticket. Want me to copy paste your notes into a new issue?

Yes, I think we should do this, even if the only thing we do with this data is log it.  I suspect that this is going to be one of the more common and banal reasons why DKG or a signing round will fail on the first try.

Reply:
And yeah this can be a separate PR :+1: Just want to make sure it's something the system can account for.